packages/ui-components/src/components/Table/FetchedTable.js

```diff
@@ -0,0 +1,129 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import React, { useCallback, useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import { getSortByKey } from '@tupaia/utils';
+import { Table } from './Table';
+import { TablePaginator } from './TablePaginator';
+import { TableHeader } from './TableHeader';
+import { TableBody } from './TableBody';
+import { tableColumnShape } from './tableColumnShape';
+
+const toggleOrder = order => (order === 'desc' ? 'asc' : 'desc');
+
+export const FetchedTable = React.memo(
+  ({
+    Header,
+    Body,
+    Paginator,
+    SubComponent,
+    columns,
+    data: initialData,
+    errorMessage,
+    noDataMessage,
+    isLoading,
+    count,
+    onChangePage,
+    onChangeRowsPerPage,
+    onRowClick,
+    orderBy: initialOrderBy,
+    order: initialOrder,
+    page,
+    rowsPerPage,
+    rowIdKey,
+    isFetching,
+    className,
+  }) => {
+    const [orderBy, setOrderBy] = useState(initialOrderBy);
+    const [order, setOrder] = useState(initialOrder);
+    const [data, setData] = useState(initialData);
+
+    useEffect(() => {
+      const ascending = order !== 'desc';
+      const newData = [...initialData].sort(getSortByKey(orderBy, { ascending }));
+      setData(newData);
+    }, [initialData, orderBy, order]);
+
+    const handleChangeOrderBy = useCallback(
+      columnKey => {
+        const newOrder = columnKey === orderBy ? toggleOrder(order) : 'asc';
+        setOrderBy(columnKey);
+        setOrder(newOrder);
+      },
+      [order, orderBy],
+    );
+
+    return (
+      <Table
+        Header={Header}
+        Body={Body}
+        Paginator={Paginator}
+        SubComponent={SubComponent}
+        columns={columns}
+        data={data}
+        errorMessage={errorMessage}
+        noDataMessage={noDataMessage}
+        isLoading={isLoading}
+        count={count}
+        onChangePage={onChangePage}
+        onChangeRowsPerPage={onChangeRowsPerPage}
+        onChangeOrderBy={handleChangeOrderBy}
+        onRowClick={onRowClick}
+        orderBy={orderBy}
+        order={order}
+        page={page}
+        rowsPerPage={rowsPerPage}
+        rowIdKey={rowIdKey}
+        isFetching={isFetching}
+        className={className}
+      />
+    );
+  },
+);
```

Comment:
```suggestion

const useSortableTable = (initialData, initialOrderBy, initialOrder) => {
  const [orderBy, setOrderBy] = useState(initialOrderBy);
  const [order, setOrder] = useState(initialOrder);
  const [data, setData] = useState(initialData);

  useEffect(() => {
    const ascending = order !== 'desc';
    const newData = [...initialData].sort(getSortByKey(orderBy, { ascending }));
    setData(newData);
  }, [initialData, orderBy, order]);

  const handleChangeOrderBy = useCallback(
    columnKey => {
      const newOrder = columnKey === orderBy ? toggleOrder(order) : 'asc';
      setOrderBy(columnKey);
      setOrder(newOrder);
    },
    [order, orderBy],
  );

  return {
    handleChangeOrderBy,
    orderBy,
    order,
    data,
  };
};

export const SortableTable = ({ data, orderBy: initialOrderBy, order: initialOrder, ...props }) => {
  const { handleChangeOrderBy, orderBy, order, data: sortedData } = useSortableTable(
    data,
    initialOrderBy,
    initialOrder,
  );
  return (
    <Table
      onChangeOrderBy={handleChangeOrderBy}
      data={sortedData}
      order={order}
      orderBy={orderBy}
      {...props}
    />
  );
};
```

Reply:
What about extracting the sorting behaviour into a custom hook and then using that to add the sorting functionality to the Table component.

- makes it clearer what is being added
- the hook could be exported and used in apps directly with tables
- don't have to keep updating all the table props. This one is debatable as I know that some people like to list out all the props but I personally think sometimes it's not worth it when we are essentially making a higher order component and using 90% of the original table functionality.

Reply:
Also what about calling it `SortableTable` rather than `FetchedTable`?

Reply:
All great points Tom! I love the idea of using a hook - I'll go ahead an implement it!

I also agree with you that in cases like this, it's fine to not list all other props. I did so here to be consistent with other similar components in this package (in specific, with the now deleted `DataFetrchingTable`). Happy to change that!

Finally, I should have added a comment: I called this component `FetchedTable` because I thought that we could also add pagination functionality (I didn't add it now because it is not required). The idea is that `FetchedTable` could have some extra features that are made possible because of the fact that its data are fetched externally, in one go.

I guess that we can implement sorting/pagination as custom hooks, and then we don't need an extra component - clients can use a combination of `Table` + `sorting hook` to achieve the desired functionality.